packages/squiggle-lang/__tests__/TS/KDE.ts

```diff
@@ -0,0 +1,59 @@
+import * as fc from "fast-check";
+import { samplesToContinuousPdf } from "../../src/rescript/Distributions/SampleSetDist/KdeLibrary.js";
+import range from "lodash/range";
+import sumBy from "lodash/sumBy";
+
+// KDE by definition, with triangular kernel max(0, |1-x|)
+const evalKde = (samples: number[], width: number, weight: number, x: number) =>
+  (weight / width) *
+  sumBy(samples, (p) => Math.max(0, 1 - Math.abs((x - p) / width)));
+
+describe("Kernel density estimation", () => {
+  test("should approximately equal naive definition", () => {
+    fc.assert(
+      fc.property(
+        fc.array(fc.double({ min: -10, max: 10, noNaN: true }), {
+          minLength: 6,
+        }),
+        fc.integer({ min: 4, max: 1e3 }),
+        fc.double({ min: 0.1, max: 3, noNaN: true }),
+        (samples, size, wantedWidth) => {
+          const len = samples.length;
+          const weight = 1 / len;
+          const sortedSamples = samples.sort((a, b) => a - b);
+          if (sortedSamples[len - 1] - sortedSamples[0] < 2 * wantedWidth)
+            return true;
+
+          const {
+            xs,
+            ys,
+            usedWidth: width,
+          } = samplesToContinuousPdf(sortedSamples, size, wantedWidth, weight);
+
+          expect(xs.length).toEqual(size);
+          expect(ys.length).toEqual(size);
+
+          // Point range should be sample range plus width on either side
+          expect(xs[0] + width).toBeCloseTo(sortedSamples[0]);
+          expect(xs[size - 1] - width).toBeCloseTo(sortedSamples[len - 1]);
+
+          // No tail
+          expect(ys[0]).toBeCloseTo(0);
+          expect(ys[size - 1]).toBeCloseTo(0);
+
+          // Checking cost is proportional to len
+          const numChecks = 10 + Math.floor(1e3 / len);
+          const checkInds =
+            numChecks >= size
```

Comment:
I'd add a 1-line comment explaining lines 47-49

Reply:
I didn't have any comment here because for me, none of this can be explained in English more clearly than it already is in the code. Here's an explanation of how I'd read this. Can you say what you're missing?

- (Background) `numChecks` is the number of checks (whatever those happen to be) to be performed.
- Ternary statement, so we'll do the first if `numChecks` is large and the second if it's small.
- The first expression is `range(size)`, which is all the indices into an array of length `size`, presumably the KDE output.
- The next step uses `Math.floor(size * Math.random())`, the standard way to get a random integer in [0, size). I don't like it for a lot of reasons and maybe we should pull it out into a utility function, but it's normal Javascript.
- Mapping over `numChecks` gives us that many of these random values. So a selection of indices into an array of length `size`.
- So we'll check every index if `numChecks` is large, and randomly-selected indices otherwise. Looking back at the condition, the threshold is that we do random selection only if the number of indices to check is smaller than the number of possible indices. That is, if we can check all indices, we do.

It's exhausting to try to guess which parts of the code would trip up an unknown reader. It would be much easier to answer specific questions about it without necessarily having the expectation that it will all be covered in comments.

Reply:
> It's exhausting to try to guess which parts of the code would trip up an unknown reader. It would be much easier to answer specific questions about it without necessarily having the expectation that it will all be covered in comments.

From my standpoint, the audience isn't me and Slava. It's the dumbest person who will need to understand this in the lifetime of the code. So specific questions from us will only go so far. 

I think my read is that with some practice, the cost of guessing what people like us will be confused by, will go down a lot.

Reply:
I feel like this code still requires me to keep more pieces of state in mind than I want. I'd probably write it a bit more like,
```
const randomIntFrom0toN = (n) =>  Math.floor(size * Math.random()))
const checkEachElement =  () => range(size)
const checkNRandomIndexesFrom0ToMax = ({n,max}) => range(n).map(randomIntFrom0toN(max)).uniq()

const indicesToCheck = 
  recommendedCheckN >= intendedOutputLen
    ? checkEachElement() 
    : checkNRandomIndexesFrom0ToMax({n: recommendedCheckN, max: intendedOutputLen)
 


```